### PR TITLE
change(rpc): add validateaddress method

### LINKED ISSUE #5722: Implement validateaddress RPC
## Motivation

Some mining pools use this RPC to check addresses.

### Specifications

https://zcash.github.io/rpc/validateaddress.html

### Designs

We can call the [`zcash_address`](https://docs.rs/zcash_address/latest/zcash_address/) crate to validate shielded and unified addresses. (Zebra also has its own transparent address parsing, but we'd be better to just be consistent and use `zcash_address` for everything.)

We just need to implement these fields:

Transparent:
- p2pkh
- p2sh
- scriptPubKey (optional, only do it if it's easy)
- pubkey (optional, only do it if it's easy)
- isscript

## Related Work

- #5719

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

Mining pools may use this RPC to check addresses.

Closes #5722.

This method returns `isvalid: false` in `zcashd` for shielded addresses, it currently checks Sapling addresses in Zebra but not Sprout, Orchard, or Unified addresses.

### Specifications

https://zcash.github.io/rpc/validateaddress.html

## Solution

- Add `zcash_address` to parse raw address
- Add `Address` enum in zebra-chain for converting `ZcashAddress`
- Add response type for validate address

## Review

Part of regular scheduled work.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- Add support for Sprout/Orchard/Unified addresses in `primitives::Address`
- Add `pubkey` and `scriptPubKey` fields

### CODE REVIEW

File: zebra-rpc/src/methods/get_block_template_rpcs.rs [hunk 1]
```diff
@@ -748,6 +758,47 @@ where
...
+                    is_valid: true,
+                    script_pub_key: None,
+                    pub_key: None,
+                    is_script: Some(address.is_script_hash()),
+                }
+            } else {
+                tracing::info!(
+                    ?network,
+                    address_network = ?address.network(),
+                    "network mismatch"
```
Comment (teor2345) [inline zebra-rpc/src/methods/get_block_template_rpcs.rs]: Can we be more specific about which addresses don't match?
```suggestion
                    "invalid address in validateaddress RPC: Zebra's configured network and address network must match"
```

File: zebra-rpc/src/methods/get_block_template_rpcs/types/validate_address.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    /// The hex encoded scriptPubKey generated by the address.
+    #[serde(rename = "scriptPubKey")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub script_pub_key: Option<HexData>,
+
+    /// The hex value of the raw public key.
+    #[serde(rename = "pubkey")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub pub_key: Option<HexData>,
```
Comment (teor2345) [inline zebra-rpc/src/methods/get_block_template_rpcs/types/validate_address.rs]: Optional Nitpick: If miners don't need these fields, can we leave them out of the code for now?

File: zebra-rpc/src/methods/tests/snapshot/get_block_template_rpcs.rs [hunk 3]
```diff
@@ -362,6 +362,18 @@ pub async fn test_responses<State, ReadState>(
...
+    let founder_address = match network {
+        Network::Mainnet => "t3fqvkzrrNaMcamkQMwAyHRjfDdM2xQvDTR",
+        Network::Testnet => "t2UNzUUx8mWBCRYPRezvA363EYXyEpHokyi",
+    };
+
+    let validate_address = get_block_template_rpc
+        .validate_address(founder_address.to_string())
+        .await
+        .expect("We should have a validate_address::Response");
+    snapshot_rpc_validateaddress(validate_address, &settings);
```
Comment (teor2345) [inline zebra-rpc/src/methods/tests/snapshot/get_block_template_rpcs.rs]: Can you add a snapshot test for an invalid address as well?

This makes sure that all the fields are skipped when they should be.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for this, I like the design here!

> This method returns `isvalid: false` in `zcashd` for shielded addresses, it currently checks Sapling addresses in Zebra but not Sprout, Orchard, or Unified addresses.

I think we might want to match `zcashd`'s behaviour for Sapling?

Otherwise nodes might try to build a transparent coinbase transaction with a Sapling address, which could cause them to fail or crash.

**Comment by teor2345** [comment]
Also there seems to be a merge conflict, can you rebase when you do your fixes?

**Comment by arya2** [comment]
> I think we might want to match zcashd's behaviour for Sapling?

I'm not sure and that's safer, so, fixed in [`fde96d5`](https://github.com/ZcashFoundation/zebra/pull/6086/commits/fde96d50573a8b4c523f6b9ceb41ec9e9ed9493d).

> Also there seems to be a merge conflict, can you rebase when you do your fixes?

Yep, I resolved those too (albeit with a merge).

File: zebra-rpc/src/methods/get_block_template_rpcs.rs (hunk 1 above)
Comment (arya2) [reply zebra-rpc/src/methods/get_block_template_rpcs.rs]: Yep, updated/applied in [`08ce203`](https://github.com/ZcashFoundation/zebra/pull/6086/commits/08ce20354dd0d9a51c4281c91b72c4617a362f4d).

File: zebra-rpc/src/methods/get_block_template_rpcs/types/validate_address.rs (hunk 2 above)
Comment (arya2) [reply zebra-rpc/src/methods/get_block_template_rpcs/types/validate_address.rs]: Okay. Removed in [`f452c89`](https://github.com/ZcashFoundation/zebra/pull/6086/commits/f452c89d645f4b3edfda0478c9460bddb2e81f38).

File: zebra-rpc/src/methods/tests/snapshot/get_block_template_rpcs.rs (hunk 3 above)
Comment (arya2) [reply zebra-rpc/src/methods/tests/snapshot/get_block_template_rpcs.rs]: Good idea. Added in [`08ce203`](https://github.com/ZcashFoundation/zebra/pull/6086/commits/08ce20354dd0d9a51c4281c91b72c4617a362f4d).

**Review by teor2345 — DISMISSED** [review]
Looks good!

**Comment by teor2345** [comment]
Can you please do a zcash-rpc-diff for this new RPC?

(Maybe we should have a testing checklist somewhere.)

**Comment by arya2** [comment]
> Can you please do a zcash-rpc-diff for this new RPC?

Yep:

<details>

```diff

Request:
validateaddress t3hgskquvKKoCtvxw86yN7q8bzwRxNgUZmc

Querying zebrad main chain at height >=1970386...

real    0m0.002s
user    0m0.002s
sys     0m0.000s

Querying zcashd main chain at height >=1830188...

real    0m0.003s
user    0m0.000s
sys     0m0.002s


Response diff between zebrad and zcashd:
--- /tmp/tmp.CZTvJsjeXg.rpc-diff/zebrad-main-1970386-validateaddress.json       2023-02-02 16:02:12.119542563 -0500
+++ /tmp/tmp.CZTvJsjeXg.rpc-diff/zcashd-main-1830188-validateaddress.json       2023-02-02 16:02:12.129542563 -0500
@@ -1,5 +1,8 @@
 {
   "isvalid": true,
   "address": "t3hgskquvKKoCtvxw86yN7q8bzwRxNgUZmc",
+  "scriptPubKey": "a914fdaa2e848fa9813b6bbc7e5f8cbd9c0724dca7d087",
+  "ismine": false,
+  "iswatchonly": false,
   "isscript": true
 }
```

```diff
Request:
validateaddress zs1knzgnu5l7h62qexv8mwv8dysr9qum9e5mtxju7hazz9qk8cpgw5uf3c3qd4n62qhdfy7q53z2c0

Querying zebrad main chain at height >=1970389...

real    0m0.002s
user    0m0.002s
sys     0m0.000s

Querying zcashd main chain at height >=1830504...

real    0m0.002s
user    0m0.002s
sys     0m0.000s


Response diff between zebrad and zcashd:
RPC responses were identical

/tmp/tmp.DmkpELZsdx.rpc-diff/zebrad-main-1970389-validateaddress.json:
{
  "isvalid": false
}
```

</details>

This brought my attention to a mismatch where Zebra was returning conversion errors.

Fixed in [`a45bff8`](https://github.com/ZcashFoundation/zebra/pull/6086/commits/a45bff8ec87fe44e7afba4560e290effdb049cc1).

**Review by teor2345 — APPROVED** [review]
Thanks for the testing!

**Pull request merged**